golang/cosmos/x/vstorage/vstorage.go

```diff
@@ -57,46 +58,90 @@ func (sh vstorageHandler) Receive(cctx *vm.ControllerContext, str string) (ret s
 	// Handle generic paths.
 	switch msg.Method {
 	case "set":
-		keeper.SetStorageAndNotify(cctx.Context, msg.Path, msg.Value)
+		for _, arg := range msg.Args {
+			var entry types.StorageEntry
+			entry, err = types.UnmarshalStorageEntry(arg)
+			if err != nil {
+				return
+			}
+			keeper.SetStorageAndNotify(cctx.Context, entry)
+		}
 		return "true", nil
 
 		// We sometimes need to use LegacySetStorageAndNotify, because the solo's
 		// chain-cosmos-sdk.js consumes legacy events for `mailbox.*` and `egress.*`.
 		// FIXME: Use just "set" and remove this case.
 	case "legacySet":
-		//fmt.Printf("giving Keeper.SetStorage(%s) %s\n", msg.Key, msg.Value)
-		keeper.LegacySetStorageAndNotify(cctx.Context, msg.Path, msg.Value)
+		for _, arg := range msg.Args {
+			var entry types.StorageEntry
+			entry, err = types.UnmarshalStorageEntry(arg)
+			if err != nil {
+				return
+			}
+			//fmt.Printf("giving Keeper.SetStorage(%s) %s\n", entry.Path(), entry.Value())
+			keeper.LegacySetStorageAndNotify(cctx.Context, entry)
+		}
 		return "true", nil
 
 	case "setWithoutNotify":
-		keeper.SetStorage(cctx.Context, msg.Path, msg.Value)
+		for _, arg := range msg.Args {
+			var entry types.StorageEntry
+			entry, err = types.UnmarshalStorageEntry(arg)
+			if err != nil {
+				return
+			}
+			keeper.SetStorage(cctx.Context, entry)
+		}
 		return "true", nil
 
 	case "append":
-		err = keeper.AppendStorageValueAndNotify(cctx.Context, msg.Path, msg.Value)
-		if err != nil {
-			return "", err
+		for _, arg := range msg.Args {
+			var entry types.StorageEntry
+			entry, err = types.UnmarshalStorageEntry(arg)
+			if err != nil {
+				return
+			}
+			if !entry.IsPresent() {
+				err = errors.New("No value for append entry with path: " + entry.Path())
+				return
+			}
+			err = keeper.AppendStorageValueAndNotify(cctx.Context, entry.Path(), entry.Value())
+			if err != nil {
+				return "", err
+			}
 		}
 		return "true", nil
 
 	case "get":
 		// Note that "get" does not (currently) unwrap a StreamCell.
-		value := keeper.GetData(cctx.Context, msg.Path)
-		if value == "" {
+		var path string
+		err = json.Unmarshal(msg.Args[0], &path)
+		if err != nil {
+			return
+		}
+
+		entry := keeper.GetData(cctx.Context, path)
+		if !entry.IsPresent() {
 			return "null", nil
 		}
-		//fmt.Printf("Keeper.GetStorage gave us %bz\n", value)
-		bz, err := json.Marshal(value)
+		//fmt.Printf("Keeper.GetStorage gave us %bz\n", entry.Value())
+		bz, err := json.Marshal(entry.Value())
 		if err != nil {
 			return "", err
 		}
 		return string(bz), nil
 
 	case "getStoreKey":
+		var path string
+		err = json.Unmarshal(msg.Args[0], &path)
+		if err != nil {
+			return
+		}
 		value := vstorageStoreKey{
 			StoreName:       keeper.GetStoreName(),
-			StoreSubkey:     string(keeper.PathToEncodedKey(msg.Path)),
+			StoreSubkey:     string(keeper.PathToEncodedKey(path)),
 			DataPrefixBytes: string(keeper.GetDataPrefix()),
+			NoDataValue:     string(keeper.GetNoDataValue()),
```

Comment:
Oof, it stinks to leak these implementation details. What would be the consequence of not exporting NoDataValue here?

Reply:
The encoding details were unfortunately already leaking to anyone that reads vstorage data directly from the chain, both in the path encoding, and in the data prefixing. It was a very unfortunate realization I couldn't get away with an internal only change. I don't fully grok how external clients like the casting package fetch data, and maybe we could have an adapter on the RPC server side to avoid leaking these details to clients? However I'd consider this out of scope for this PR. Ping @michaelfig

Reply:
The only way an external client can verify a Merkle proof for some on-chain data from an RPC node (trusting the validators but not the integrity of the RPC node) is by querying the exact IAVL key/value representation in bytes, and then knowing how they correspond to the application layer.  This verification process necessarily crosses abstractions in an uncomfortable way.

Reply:
If you're proposing to remove `getStoreKey` and nothing in the SDK uses it, that's fine with me.

Reply:
> This verification process necessarily crosses abstractions in an uncomfortable way.

Makes sense

> If you're proposing to remove `getStoreKey` and nothing in the SDK uses it, that's fine with me.

I remember seeing some potential uses. I didn't feel comfortable enough changing that wiring.